### PR TITLE
 Automatically calculate perplexity metrics for model supportability

### PR DESCRIPTION
**Pull request by ayissi-msft**
This PR aims to enhance our model validation process by automatically calculating the perplexity metrics if the model has the corresponding dataset. This will allow us to generate a strong report on the model's supportability, providing valuable insights into its performance and reliability.

### CODE REVIEW

File: tools/python/model_validation/perplexity_metrics.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+        params = og.GeneratorParams(model)
+        generator = og.Generator(model, params)
+
+        logits = generator.compute_logits("logits")
+
+        targets = np.roll(input_ids, -1, axis=1)
+
+        # Use LogSoftMax here
+        log_probs = torch.nn.functional.log_softmax(torch.tensor(logits), dim=1).numpy()
```
Comment (fajin-corp) [inline tools/python/model_validation/perplexity_metrics.py:32]: >1 [](http://example.com/codeflow?start=78&length=1)

probably -1

File: tools/python/model_validation/perplexity_metrics.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+        batch_size, seq_length = targets.shape
+        target_log_probs = log_probs[np.arange(batch_size)[:, None], np.arange(seq_length), targets]
```
Comment (fajin-corp) [inline tools/python/model_validation/perplexity_metrics.py:35]: >target_log_probs = log_probs[np.arange(batch_size)[:, None], np.arange(seq_length), targets] [](http://example.com/codeflow?start=8&length=92)

this is the critical part

File: tools/python/model_validation/perplexity_metrics.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+def perplexity_eval(model_dir):
+    model = og.Model(f'{model_dir}')
+    tokenizer = og.Tokenizer(model)
+
+    dataset = get_wikitext2()
+
+    total_log_probs = 0
+    total_token_count = 0
+
+    for batch in dataset:
```
Comment (fajin-corp) [inline tools/python/model_validation/perplexity_metrics.py:19]: >batch in dataset [](http://example.com/codeflow?start=8&length=16)

concatenate all batches

File: tools/python/model_validation/perplexity_metrics.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        text = batch["text"]
+
+        input_ids = tokenizer.encode_batch([text])
+
+        params = og.GeneratorParams(model)
+        generator = og.Generator(model, params)
+
+        logits = generator.compute_logits("logits")
+
+        targets = np.roll(input_ids, -1, axis=1)
```
Comment (fajin-corp) [inline tools/python/model_validation/perplexity_metrics.py:29]: >np.roll(input_ids, -1, axis=1) [](http://example.com/codeflow?start=18&length=30)

what does it do?

**Review by fajin-corp — APPROVED** [review]
